Plugin system restyling

NeroReflex opened the pull request:
Cambiamenti **STRUTTURALI**:
- i plugin tornano a NON avere tipo, quindi un plugin unisce entrambe le tipologie.
- il parsing del messaggio va fatto a mano (anche se il vecchio parsing rimane disponibile per retro-compatibilità e perché può far comodo ai neofiti)
- ogni plugin lancia ben DUE Threads (per evitare che i due meccanismi di plugin si limitino fra loro)
- le funzioni si lanciano così:
  !convert hex2bin 1aff5
- una facile ed intuitiva guida si invoca così: !help
- la guida di uno specifico plugin si invoca così: !help convert
- maggiore velocità a runtime, soprattutto nel main thread (parte critica in Pizza)

NeroReflex committed:
Cambiamento nella struttura del progetto

Ora un plugin NON ha più un tipo e ha un'unica coda di richieste.
Ogni richiesta ha un tipo (in sostituzione del tipo di plugin).
Ogni plugin lancia (al massimo) DUE thread (uno per tipo di richiesta).
Cambiata l'interfaccia per rispondere alle richieste.

NeroReflex committed:
Migliore usabilità

doppioandante commented on src/main/java/com/neroreflex/pizza/Pizza.java:
```diff
@@ -91,79 +87,77 @@ public final String getBotID() {
+            } else if (command.compareTo("help") == 0) {// E se non lo e' invia l'errore
+                this.Help(new Request(channel, sender, rawParamsString));
+            } else {
+                this.enqueueMessage(new Message(channel, "I don't know what I have to do at '" + command + "' request :("));
```
Nel caso il comando non esista secondo me è più chiaro non scrivere nulla.

NeroReflex replied:
!time
dopo un ipotetico secondo.....
ma è morto o il plugin non esiste?

doppioandante replied:
!help dà la risposta

NeroReflex replied:
Non lo so.... Mi sembra brutto (e un po' maleducato)....

doppioandante commented on src/main/java/com/neroreflex/pizza/Pizza.java:
```diff
@@ -91,79 +87,77 @@ public final String getBotID() {
-                if (args.elementAt(0).compareTo("list") == 0) {
-                    this.tranci.entrySet().stream().forEach((entry) -> {
+        } else if (message.substring(0, nickLength).compareTo(this.getNick()) == 0) {
+            this.enqueueMessage(new Message(channel, "Sono il vostro amichevole robottino mangiapizza :)"));
```
Sincramente toglierei questo messaggio, a parte che non fa simpatia è pure in Italiano mentre gli altri sono in Inglese.

NeroReflex replied:
Per questo mi piace :D

doppioandante commented on src/main/java/com/neroreflex/pizza/Pizza.java:
```diff
@@ -91,79 +87,77 @@ public final String getBotID() {
-                    this.enqueueMessage(new Message(channel, "The requested plugin cannot be actively activated"));
-                }
-            } // E se non lo e' invia l'errore
+        } // Devo stampare SOLO l'help del plugin richiesto
```
I commenti come questo sono un po' lapalissiani, mi pare che il codice si capisca bene anche senza.

doppioandante commented on src/main/java/com/neroreflex/pizza/Pizza.java:
```diff
@@ -91,79 +87,77 @@ public final String getBotID() {
+            } // Se non lo ho avverto l'utente
             else {
-                this.enqueueMessage(new Message(channel, "I don't know what I have to do at '" + command + "' request :("));
+                this.enqueueMessage(new Message(helpRequest.GetChannel(), "A plugin named '" + name + "' doesn't exist."));
```
No plugin named 'name' exists

doppioandante commented on src/main/java/com/neroreflex/pizza/Pizza.java:
```diff
@@ -91,79 +87,77 @@ public final String getBotID() {
+                if (!help.isEmpty())
+                    this.enqueueMessage(new Message(helpRequest.GetChannel(), "!" + name + " " + help));
+                else
+                    this.enqueueMessage(new Message(helpRequest.GetChannel(), "Usage help is not bundled with the '" + name + "' plugin."));
```
Help e basta invece di Usage help

doppioandante commented on src/main/java/com/neroreflex/pizza/Request.java:
```diff
@@ -27,24 +27,34 @@
+        Vector<String> args = new Vector<>();
+        
+        for (int i = 0; i < params.length; i++)
+            if (params[i].length() > 0) args.add(params[i]);
```
Non dovrebbe essere impossibile che i capture abbiano lunghezza nulla?

NeroReflex replied:
Secondo i test condotti da me e GN direi proprio di no (purtroppo)!

doppioandante commented on src/main/java/com/neroreflex/pizza/Trancio.java:
```diff
@@ -126,24 +243,78 @@ protected void finalize() throws Throwable {
     /*      BUON DIVERTIMENTO      */
     
-    protected /*abstract*/ void onInitialize()/*;*/ {}
+    protected void onInitialize() {}
```
Perché non sono abstract?

NeroReflex replied:
se fossero abstract saresti costretto a definirli. E' abstract la classe, può bastare.

doppioandante commented:
Ci ho dato uno sguardo molto veloce, una cosa che mi preme è che con @gianluca-nitti avevamo fatto in modo che il bot di hackernews andasse a intervalli regolari ma a partire da ore prestabilite, cosa che si è persa quando hai messo l'autotrancio.
Non ho ancora guardato cosa intendi per 'ogni plugin lancia ben DUE Threads' ma a sentirla così non sembra una grande cosa, guarderò meglio se avrò tempo.

## Risposte

La possibilità di far partire l'onPoll a intervalli regalari a partire da ore prestabilite è stata introdotta per tutti i plugin.

DUE Thread: uno per il timer e uno per la blockingQueue che gestisce l'onCall. Avere un solo thread significherebbe accodare onCall e onPoll o spendere più tempo a runtime per capire quale eseguire,
ma come è ora è più leggibile.

NeroReflex committed:
I thread vengono eseguiti PRIMA di settare loaded

I thread vengono eseguiti PRIMA di settare loaded, quindi isLoaded()
ritorna un valore dipendente dal tempo di scarto tra inizio esecuzione
thread e fine caricamento plugin. In pratica un plugin funzionerà se e
solo se ha fortuna!

NeroReflex committed:
I commenti come questo sono un po' lapalissiani.

NeroReflex committed:
Recuperato codice da HackerNews e reso API

Implementato un modo per definere la prima chiamata ad onPoll.
Il codice precedentemente usato per il plugin HackerNews è tornato, e
funge anche da esempio per gli altri!

gianluca-nitti commented:
@NeroReflex @doppioandante c'è qualche motivo particolare per cui non è ancora stato fatto il merge in master?

doppioandante commented:
Perché @NeroReflex aveva trovato un problema che adesso non ricordo.
A me sinceramente non piace il discorso dei due thread, dovrebbero essercene due solo se il plugin richiede l'attivazione periodica.
E comunque ancora non mi convince del tutto che ci sia questo kitchen sink dentro il plugin, preferivo una soluzione come quella dell'autotrancio che aggiunge la funzionalità del polling estendendo la classe Trancio: prima l'architettura del tutto rendeva le due cose esclusive, ora invece si può benissimo implementare così.

Un'altra cosa che non ho capito a che scopo sia stata messa è DelayedRequest.

A parte questo non ho molto altro in contrario. In futuro mi piacerebbe che Request e Message venissero unificati perché non ha senso avere una distinzione semantica riguardo alla provenienza, è già implicita nella gerarchia di comunicazione bot/plugin.

NeroReflex committed:
Rimosso il messaggio di cattivo gusto

NeroReflex committed:
Non più necessario

doppioandante commented on src/main/java/com/neroreflex/pizza/Pizza.java:
```diff
@@ -91,79 +87,91 @@ public final String getBotID() {
-        Matcher invokeMatcher = invokeRegex.matcher(message);
-        if (invokeMatcher.matches()) {
+        // Cerco di capire se mi e' stato dato un comando
+        if ((message.charAt(0) == '!') && (message.length() > 1)) {
```
Può essere che arrivi un messaggio di lunghezza nulla?

NeroReflex replied:
https://tools.ietf.org/html/rfc2812.html Sezione 2.3.1: "Empty messages are silently ignored"

doppioandante commented on src/main/java/com/neroreflex/pizza/Pizza.java:
```diff
@@ -91,79 +87,91 @@ public final String getBotID() {
             // Ottieni il nome del comando
-            String command = invokeMatcher.group(3).toLowerCase();
+            int firstSpacePos = message.indexOf(' ');
+            firstSpacePos = (firstSpacePos <= -1)? message.length() : firstSpacePos;
```
Nit: di solito ci va lo spazio prima del ? del ternario, ma lascia pure così.

doppioandante commented on src/main/java/com/neroreflex/pizza/Pizza.java:
```diff
@@ -91,79 +87,91 @@ public final String getBotID() {
+            } else if (command.compareTo("info") == 0) {// E se non lo e' invia l'errore
+                this.Info(new Request(channel, sender, rawParamsString));
+            } else {
+                this.enqueueMessage(new Message(channel, "I don't know what I have to do at '" + command + "' request :("));
```
La grammatica corretta sarebbe "to do with", e probabilmente quel "request" alla fine va tolto.

doppioandante commented on src/main/java/com/neroreflex/pizza/Pizza.java:
```diff
@@ -91,79 +87,91 @@ public final String getBotID() {
+            this.enqueueMessage(new Message(helpRequest.GetChannel(), "My source code <3 https://github.com/NeroReflex/Pizza"));
+        } else {
+            this.enqueueMessage(new Message(helpRequest.GetChannel(), "No info about that"));
+        }
```
Penso che sarebbe più semplice far stampare semplicemente I'm blabla + url github (com'era prima forse?), è una complicazione inutile avere dei subcommands per info.

doppioandante commented on src/main/java/com/neroreflex/pizza/Pizza.java:
```diff
@@ -91,79 +87,91 @@ public final String getBotID() {
+                        
+                        // Stampo la guida solo se ha senso (la stringa della guida NON E' vuota)
+                        if (!help.isEmpty())
+                            this.enqueueMessage(new Message(helpRequest.GetChannel(), "!" + name + " " + help));
```
Per ora OK, ma il si dovrà cambiare in modo che il messaggio con le informazioni di help arrivi in privato all'utente che l'ha richiesto, altrimenti intasa la chat (specie con molti plugin o con help lunghi).

doppioandante commented on src/main/java/com/neroreflex/pizza/Pizza.java:
```diff
@@ -91,79 +87,91 @@ public final String getBotID() {
+                if (!help.isEmpty())
+                    this.enqueueMessage(new Message(helpRequest.GetChannel(), "!" + name + " " + help));
+                else
+                    this.enqueueMessage(new Message(helpRequest.GetChannel(), "Help is not bundled with the '" + name + "' plugin."));
```
(There's) No help is provided by <name>

doppioandante commented on src/main/java/com/neroreflex/pizza/Trancio.java:
```diff
@@ -5,18 +5,124 @@
+            throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
+        }
+        
+    };
```
RequestRunner e TimerRunner si potevano fare come classi anonime ma va bene anche così.

doppioandante commented on src/main/java/com/neroreflex/plugins/Join.java:
```diff
@@ -17,42 +17,37 @@
+            this.sendMessage(new Message(channel, user + " I have joined the " + newChannel + " channel"));
         } else {
-            this.sendMessage(new Message(channel, user + "Please specify a channel"));
+            this.sendMessage(new Message(channel, user + " Invalid channel name"));
```
Questo else avviene quando non si dà nessun canale, quindi ha più senso il messaggio che c'era prima.

doppioandante commented on src/main/java/com/neroreflex/plugins/Join.java:
```diff
@@ -17,42 +17,37 @@
 
             this.joinChannel(newChannel, key);
-            this.sendMessage(new Message(channel, user + " I have joined " + newChannel));
+            this.sendMessage(new Message(channel, user + " I have joined the " + newChannel + " channel"));
```
Il messaggio di prima era meno verboso e si capisce ugualmente per via del # prima del nome del canale.

NeroReflex committed:
L'help risponde in privato.

NeroReflex committed:
HackerNews torna (meglio) di come era prima
